### PR TITLE
[FIX] hr_recruitment, web: the sidepanel doesn't show users correctly

### PR DESCRIPTION
**Pull request by federico-galfione**
How to replicate:
- Create a user and assign the current company to its
related contact.
- Go to Recruitment -> create a job with the user
created before as Responsible.

Result:
- The user doesn't display in the sidepanel.

Expected:
- The panel should display the user and the other
users that are responsible of at least one job.

Additionally, if the user's contact has no company
assigned, it will display in the panel whether or not
the user is set as responsible in at least one job.

I overwrote the 'search_panel_select_range' to
display only the users that respect the rules listed
above.
I also changed the original search_panel_select_range
and now I don't set parent_field if the parent_field
model is different than the current model

opw-2654573

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by nle-odoo — COMMENTED** [review]
Hi

```
A RecruitmentSearchPanel widget was created as an
extension of SearchPanel to refresh the categories
each time the panel is displayed.
```

The code here is modifying web.SearchPanel widget and modifying it (so it will be applied to all search panels).

I don't understand why a refresh is needed. Will the call (to search_panel_select_range) not always be the same and always give us the same return?

For the override of `search_panel_select_range` it looks alright but  this:

```py
            available_users = [
                {'id': user.id, 'display_name': user.name} 
                for user in self.env['res.users'].search([]) 
                if user in [job.user_id for job in self.env['hr.job'].search([]) if job.user_id]
            ]
```

could be changed to use more search/search_read that decrease SQL queries, like this I think:

```py
users = self.env['hr.job'].search([('user_id', '!=', False)]).mapped('user_id')
available_users = self.env['res.users'].search([('id', 'in', users.ids)], ['display_name'], limit=SEARCH_PANEL_LIMIT)
```

As a side note, the original issue (res.users with parent_id that is not a res.users) seems to happen because in this code:

https://github.com/odoo/odoo/blob/303d3be40c6a24a19da00a7a406206bd09d08337/addons/web/static/src/js/views/search_panel.js#L295-L315

where the result with parent are removed and added as child to their parent.

but if the parent is not there (which is the case here, since the parent is not a res.users but a res.partner, as a side note if a partner had the same ID as a shown user this could lead to even more issue) they will just be removed and never shown. I think that issue should be solved:

- possibly there
- or even in the original `search_panel_select_range` we should not give `parent_field` if the parent_field model is different than the current model (so no if parent is res.partner and field is res.users, it should totally ignore parent by default)

For the report that you see users without any job application, that is nice to have it (with a change in python `def search_panel_select_range` as done here) but I'm not sure this is necessary.

__so in summary:__

1. it would be good to solve the real issue about record not being shown
2. if we do a change to hide users that are not responsible this would be a nice to have (and in 14.0 the search panel seems to be removed so it would be adding code for a nice to have to something that is already removed)

**Comment by federico-galfione** [comment]
> I don't understand why a refresh is needed. Will the call (to search_panel_select_range) not always be the same and always give us the same return?

The refresh is needed in the case we keep the "hide users that are not responsible" behavior. If so, if you enter in recruitment you'll see a list of users, but if you edit a job, change the responsible and come back to the jobs page, the sidebar has to be refreshed to show the new list of users responsible of at least one job.
If this is not necessary, the refresh would be no longer useful and we can get rid of it.

File: addons/hr_recruitment/models/hr_job.py [hunk 1]
```diff
@@ -108,6 +108,17 @@ def create(self, vals):
         vals['favorite_user_ids'] = vals.get('favorite_user_ids', []) + [(4, self.env.uid)]
         return super(Job, self).create(vals)
 
+    @api.model
+    def search_panel_select_range(self, field_name):
+        if field_name == 'user_id':
+            users = self.env['hr.job'].search([('user_id', '!=', None)]).mapped('user_id')
+            available_users = self.env['res.users'].search([('id', 'in', users.ids)], limit=200).mapped(lambda user: { 'id': user.id, 'display_name': user.name})
```
Comment (nle-odoo) [inline addons/hr_recruitment/models/hr_job.py:115]: instead of `200` you should use:

```py
from odoo.addons.web.models.models import SEARCH_PANEL_LIMIT
```
instead of:

```py
.mapped(lambda user: { 'id': user.id, 'display_name': user.name})
```

I think you can do:

```py
.read(['display_name'])
```

File: addons/web/models/models.py [hunk 2]
```diff
@@ -258,7 +258,7 @@ def search_panel_select_range(self, field_name):
 
         Comodel = self.env[field.comodel_name]
         fields = ['display_name']
-        parent_name = Comodel._parent_name if Comodel._parent_name in Comodel._fields else False
+        parent_name = Comodel._parent_name if Comodel._parent_name in Comodel._fields and Comodel._fields[Comodel._parent_name].comodel_name == Comodel._name else False
```
Comment (nle-odoo) [inline addons/web/models/models.py:261]: I think this could be simplified with:

```py
parent_name = Comodel._fields.get(Comodel._parent_name).comodel_name == Comodel._name
```

**Review by nle-odoo — COMMENTED** [review]
Hi,


After testing a little, if you have:

userID1: {partner_id: partnerID1, parent_id: partnerID2}
userID2: {partner_id: partnerID4, parent_id: partnerID5}
userID3: {partner_id: partnerID2, parent_id: partnerID6}

if you search:

`[('id', 'child_of', userID2)]`

you will get userID2 but also userID1 because we search on parent_id as if it was a res.users and not a res.partner.

I'm not sure of the best fix for that:

- code like here
- have code in res.users similar to "_parent_name = False"

we should probably check with ORM (eg. rco).

In any case the first part of the fix in hr.job would fix the particular issue of the customer.

File: addons/hr_recruitment/models/hr_job.py [hunk 3]
```diff
@@ -108,6 +109,17 @@ def create(self, vals):
         vals['favorite_user_ids'] = vals.get('favorite_user_ids', []) + [(4, self.env.uid)]
         return super(Job, self).create(vals)
 
+    @api.model
+    def search_panel_select_range(self, field_name):
+        if field_name == 'user_id':
+            users = self.env['hr.job'].search([('user_id', '!=', None)]).mapped('user_id')
+            available_users = self.env['res.users'].search([('id', 'in', users.ids)], limit=SEARCH_PANEL_LIMIT).read(['display_name'])
```
Comment (rco-odoo) [inline addons/hr_recruitment/models/hr_job.py:116]: This might be even more readable
```suggestion
            users = self.search([('user_id', '!=', False)]).user_id
            available_users = users._filter_access_rules('read')[:SEARCH_PANEL_LIMIT].read(['display_name'])
```

File: addons/web/models/models.py [hunk 4]
```diff
@@ -258,7 +258,8 @@ def search_panel_select_range(self, field_name):
 
         Comodel = self.env[field.comodel_name]
         fields = ['display_name']
-        parent_name = Comodel._parent_name if Comodel._parent_name in Comodel._fields else False
+        parent_fields = Comodel._fields.get(Comodel._parent_name)
+        parent_name = Comodel._parent_name if parent_fields and parent_fields.comodel_name == Comodel._name else False
```
Comment (rco-odoo) [inline addons/web/models/models.py:262]:
```suggestion
        parent_field = Comodel._fields.get(Comodel._parent_name)
        parent_name = Comodel._parent_name if parent_field and parent_field.comodel_name == Comodel._name else False
```

**Review by rco-odoo — COMMENTED** [review]
Looks good to me.
Let's avoid changing `_parent_name` in the ORM if possible.

**Review by nle-odoo — APPROVED** [review]
i'm ok to merge

just a minor detail, I think we should have back:

```diff
- available_users = users._filter_access_rules('read')[:SEARCH_PANEL_LIMIT].read(['display_name'])
+ available_users = users.search([('id', 'in', users.ids)], limit=SEARCH_PANEL_LIMIT).read(['display_name'])
```

so that the records are sorted (as originally and in other search panel), and if we have more than the limit, the cut can be noticed more easily (eg. all records after P are not shown).

**Comment by nle-odoo** [comment]
robodoo r+
